Adds new SavedObjectsRespository error type for 404 that do not originate from Elasticsearch responses

TinaHeiligers opened the pull request:
## Summary
Resolves https://github.com/elastic/kibana/issues/102353.

Replaces https://github.com/elastic/kibana/pull/107104.

There [may be cases](https://github.com/elastic/kibana/blob/036c157f10081a39ba0a260f191c37794ed90a2a/src/core/server/saved_objects/service/lib/repository.ts#L1031-L1040) where a 404 can be returned during calls to the Saved Objects Repository for reasons other than the saved object not being found. 

The Saved Objects service should distinguish between a Saved Object is not found and other unknown reasons that do not originate from responses to an Elasticsearch call.

This PR adds an explicit 503 NotFoundEsUnavailable Error that will wrap 404 errors from `get`, `update` and `delete` requests when the response does not contain the `x-elastic-product: Elasticsearch` header.

As a follow up to this initial implementation, we need to:
- Add general product check to all responses from the SO repository calls to ES https://github.com/elastic/kibana/issues/107343.
- Add an integration test that uses a proxy to mimic the behavior of the client when the response does not pass the product check.https://github.com/elastic/kibana/issues/107246
- Handle document not found and index not found response conditions consistently https://github.com/elastic/kibana/issues/107944

### Checklist

Delete any items that are not applicable to this PR.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

### Risk Matrix

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| False positives&mdash; An intermediate proxy doesn't forward `x-elastic-product` header . | Low | Low | As of 7.15.0 Kibana won't start unless the required headers are present. For 7.14 patches, the type of error returned could be a 404 or a 503 but we will still get an error|

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/141994/)
* Commit: 98a785fe5a8fac01e138e82d294686021a416e38
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Check Doc API Changes](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/141994/execution/node/425/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/141994/","number":"141994","commit":"98a785fe5a8fac01e138e82d294686021a416e38"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142016/)
* Commit: e173b778df1e5d5b00c19b1e1c6c6144ff9002d9
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-107301/e173b778df1e5d5b00c19b1e1c6c6144ff9002d9)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/121bccb4d3eaa8e1cd17b2a75b3afcd765b86cc8) | [after](https://github.com/elastic/kibana/commit/e173b778df1e5d5b00c19b1e1c6c6144ff9002d9) | diff |
| --- | --- | --- | --- |
| `core` | 2363 | 2368 | +5 |

#### API count missing comments
| id | [before](https://github.com/elastic/kibana/commit/121bccb4d3eaa8e1cd17b2a75b3afcd765b86cc8) | [after](https://github.com/elastic/kibana/commit/e173b778df1e5d5b00c19b1e1c6c6144ff9002d9) | diff |
| --- | --- | --- | --- |
| `core` | 1081 | 1086 | +5 |


</details>

### History
* :broken_heart: [Build #141994](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/141994/) failed 98a785fe5a8fac01e138e82d294686021a416e38

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/141994/","number":"141994","commit":"98a785fe5a8fac01e138e82d294686021a416e38"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142016/","number":"142016","commit":"e173b778df1e5d5b00c19b1e1c6c6144ff9002d9"}]}
PIPELINE-->

TinaHeiligers commented on src/core/server/saved_objects/service/lib/errors.ts:
```diff
@@ -202,4 +202,23 @@ export class SavedObjectsErrorHelpers {
+    const notFoundError = this.createGenericNotFoundError(type, id);
+    return this.decorateEsUnavailableError(
+      new Error(`${notFoundError.message}`),
+      `x-elastic-product not present or not recognized`
```
Copied self-review comment:

> I'm not sure this is the best text string to use as a descriptor for the "missing" header. I took [inspiration](https://github.com/elastic/elasticsearch-js/blob/master/lib/errors.js?rgh-link-date=2021-07-29T00%3A11%3A30Z#L141) from the client's team but made it a lot more specific. We need to get Product's input on wording if we choose to follow this implementation.

TinaHeiligers commented on src/core/server/saved_objects/service/lib/errors.ts:
```diff
@@ -202,4 +202,23 @@ export class SavedObjectsErrorHelpers {
+  public static isNotFoundEsUnavailableError(error: Error | DecoratedError) {
+    return (
+      isSavedObjectsClientError(error) &&
+      error[code] === CODE_ES_UNAVAILABLE &&
```
Copied self-review comment:

> Specifically set a status code of 503 to indicate an ES availability error. This overrides the 404 we would otherwise have thrown.

TinaHeiligers commented on src/core/server/saved_objects/service/lib/repository.ts:
```diff
@@ -665,9 +665,15 @@ export class SavedObjectsRepository {
+        throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+      } else {
+        // throw if we can't verify the response is from Elasticsearch
+        throw SavedObjectsErrorHelpers.createGenericNotFoundEsUnavailableError(type, id);
```
Self-review:
`createGenericNotFoundEsUnavailableError` will return a [503](https://github.com/elastic/kibana/pull/107104/files#diff-c6ac0127a77f2ba335589ef96b96203ca6b393d22a1878e95225c912c0e86d51R489) error that adds [the message from a `NotFoundError`](https://github.com/elastic/kibana/pull/107104/files#diff-c6ac0127a77f2ba335589ef96b96203ca6b393d22a1878e95225c912c0e86d51R483) to the `EsUnavailableError`. The decorated error that's thrown will [assert `true` to a check for `isEsUnavailableError`](https://github.com/elastic/kibana/blob/master/src/core/server/saved_objects/service/lib/errors.ts#L195), since the latter returns a `boolean` for the `statusCode` being 503.
If a saved object type and id are provided, the error will have the form:
```
{
	message: 'x-elastic-product not present or not recognized: Saved object [foo/bar] not 	found',
	statusCode: 503
}
```
Otherwise, the error will have the form:
```
{
	message: 'x-elastic-product not present or not recognized: Not Found',
	statusCode: 503
}
```

mshustov replied:
optional nit: I had to read `createGenericNotFoundEsUnavailableError` implementation to understand what error it will create. Maybe we should pass an error reason to `decorateEsUnavailableError` explicitly?
```js
decorateEsUnavailableError(error, `Saved object [${type}/${id}] not found`);
```

> // throw if we can't verify the response is from Elasticsearch


Will we add such a helper for every possible error response?

TinaHeiligers replied:
@joshdover which question did you give the thumbs up for in https://github.com/elastic/kibana/pull/107301#discussion_r682480245? i.e.

> Maybe we should pass an error reason to decorateEsUnavailableError explicitly?

or 

> Will we add such a helper for every possible error response?

I'm not sure which way we want to go eventually, as long as we provide enough info to easily figure out the root cause of an error when it's thrown.

TinaHeiligers replied:
> Maybe we should pass an error reason to decorateEsUnavailableError explicitly?

We would still need to provide an error to pass to `decorateEsUnavailableError`:
```
public static decorateEsUnavailableError(error: Error, reason?: string) {
    return decorate(error, CODE_ES_UNAVAILABLE, 503, reason);
  }
```
which is what we're doing in `createGenericNotFoundEsUnavailableError`.

TinaHeiligers reviewed (commented):
Comments to reviewers.

TinaHeiligers reviewed (commented):
Code comments.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142074/)
* Commit: 5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-107301/5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/121bccb4d3eaa8e1cd17b2a75b3afcd765b86cc8) | [after](https://github.com/elastic/kibana/commit/5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b) | diff |
| --- | --- | --- | --- |
| `core` | 2363 | 2368 | +5 |

#### API count missing comments
| id | [before](https://github.com/elastic/kibana/commit/121bccb4d3eaa8e1cd17b2a75b3afcd765b86cc8) | [after](https://github.com/elastic/kibana/commit/5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b) | diff |
| --- | --- | --- | --- |
| `core` | 1081 | 1086 | +5 |


</details>

### History
* :green_heart: [Build #142016](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142016/) succeeded e173b778df1e5d5b00c19b1e1c6c6144ff9002d9
* :broken_heart: [Build #141994](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/141994/) failed 98a785fe5a8fac01e138e82d294686021a416e38

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/141994/","number":"141994","commit":"98a785fe5a8fac01e138e82d294686021a416e38"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142016/","number":"142016","commit":"e173b778df1e5d5b00c19b1e1c6c6144ff9002d9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142074/","number":"142074","commit":"5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b"}]}
PIPELINE-->

TinaHeiligers commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142325/)
* Commit: 7d00fdd124fb23ec86e99e206fe754ac1655f8fc
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-107301/7d00fdd124fb23ec86e99e206fe754ac1655f8fc)
* [Documentation Changes](https://kibana_107301.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/df421f869feb747256bfa329efbbf8c30c603e0e) | [after](https://github.com/elastic/kibana/commit/7d00fdd124fb23ec86e99e206fe754ac1655f8fc) | diff |
| --- | --- | --- | --- |
| `core` | 2363 | 2368 | +5 |

#### API count missing comments
| id | [before](https://github.com/elastic/kibana/commit/df421f869feb747256bfa329efbbf8c30c603e0e) | [after](https://github.com/elastic/kibana/commit/7d00fdd124fb23ec86e99e206fe754ac1655f8fc) | diff |
| --- | --- | --- | --- |
| `core` | 1081 | 1086 | +5 |


</details>

### History
* :green_heart: [Build #142074](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142074/) succeeded 5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b
* :green_heart: [Build #142016](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142016/) succeeded e173b778df1e5d5b00c19b1e1c6c6144ff9002d9
* :broken_heart: [Build #141994](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/141994/) failed 98a785fe5a8fac01e138e82d294686021a416e38

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/141994/","number":"141994","commit":"98a785fe5a8fac01e138e82d294686021a416e38"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142016/","number":"142016","commit":"e173b778df1e5d5b00c19b1e1c6c6144ff9002d9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142074/","number":"142074","commit":"5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142325/","number":"142325","commit":"7d00fdd124fb23ec86e99e206fe754ac1655f8fc"}]}
PIPELINE-->

YulNaumenko reviewed (approved):
LGTM! Awesome work! I've tested this locally by simulating situation with ES restart during the Kibana was still running - as a result all scheduled rules continue running after Kibana got back to the healthy state.
This PR should resolve this Alerting [issue](https://github.com/elastic/kibana/issues/100764) with no actions from the our side

joshdover commented:
@mshustov regarding [your comment](https://github.com/elastic/kibana/pull/107104#discussion_r679113717) regarding using this header before our version check uses it:

> Question: Might it lead to false positives if an intermediate proxy doesn't forward `x-elastic-product` header? In `v7.14`, we do allow Kibana starts even if `x-elastic-product` header is missing. Kibana won't start unless the header presents starting from `v7.15` https://github.com/elastic/kibana/issues/105557
Does it mean that we cannot backport this change to `v7.14.1`?

In addition to @TinaHeiligers's note that this check only applies to 404, it also seems quite possible that we will have #105557 fixed for the 7.14.1 release which makes me less concerned about this. Though maybe we should consider delaying a backport until that change is also backported.

TinaHeiligers commented:
> Though maybe we should consider delaying a backport until that change is also backported

@joshdover do you mean the 7.14.1 backport or both backports to 7.14.1 and 7.x (7.15)?

joshdover commented on src/core/server/elasticsearch/client/mocks.ts:
```diff
@@ -139,15 +139,16 @@ export type MockedTransportRequestPromise<T> = TransportRequestPromise<T> & {
   body: T,
-  { statusCode = 200 }: { statusCode?: number } = {}
+  { statusCode = 200 }: { statusCode?: number } = {},
+  headers?: U
 ): MockedTransportRequestPromise<ApiResponse<T>> => {
-  const response = createApiResponse({ body, statusCode });
+  const response = createApiResponse({ body, statusCode, headers });
   const promise = Promise.resolve(response);
-  (promise as MockedTransportRequestPromise<ApiResponse<T>>).abort = jest.fn();
+  (promise as MockedTransportRequestPromise<ApiResponse<T, U>>).abort = jest.fn();
 
-  return promise as MockedTransportRequestPromise<ApiResponse<T>>;
+  return promise as MockedTransportRequestPromise<ApiResponse<T, U>>;
```
I'm not sure I'm following the type change for the new `U` type param here. The second type parameter to `ApiResponse` is for the `meta.context` field and I don't see it applied to the `headers` field at all. I think this may be unnecessary?

TinaHeiligers replied:
@joshdover It's the only way I could actually get the headers added to the mock. If there's a better way, please let me know! I'm by no means a typescript expert.

TinaHeiligers replied:
done.

joshdover commented on src/core/server/saved_objects/service/lib/errors.ts:
```diff
@@ -202,4 +202,23 @@ export class SavedObjectsErrorHelpers {
+    );
+  }
+
+  public static isNotFoundEsUnavailableError(error: Error | DecoratedError) {
```
Do we need this API? If we want plugins to treat this error as any other ES not found issue, I think it'd be ok to only include the factory function `createGenericNotFoundEsUnavailableError`  so that plugins don't waste time trying to figure out how they should handle this error in a different way.

TinaHeiligers replied:
I'm actually using the API within the repository to verify error types [here](https://github.com/elastic/kibana/pull/107301/files?file-filters%5B%5D=.js&file-filters%5B%5D=.ts#diff-01c073903639e29be8562ed0de79fae4084a953304672d1e6bdb3fd7fef06f11R1266). It also ensures we stay consistent with repository error type checks, such as [`isNotFoundError`](https://github.com/elastic/kibana/pull/107301/files?file-filters%5B%5D=.js&file-filters%5B%5D=.ts#diff-7a19c0f1ba03e2b736d8031dc6c5ea8b746cae37d2d81f357577ec34612dbc12R151-R153), with the additional check on the error reason.

joshdover replied:
Right, but as a plugin developer, I'd want to handle each error exposed on `SavedObjectsErrorHelpers.is*` in order to make my code bulletproof. I'd expect that each of these `is*Error` methods to only correspond to one error type and not overlap, but now that is no longer the case since both `isEsUnavailableError` and `isNotFoundEsUnavailableError` could return true on the same error. I'd also expect that each error type would correspond to something different I could do to handle this, however the underlying problem for both of these errors is the same and there's really nothing different I can/should do as a developer for the `isNotFoundEsUnavailableError` case vs the `isEsUnavailable` one.

Are you planning to add similar new error guards for each method as part of https://github.com/elastic/kibana/issues/107343? Personally, I think we should only expose `isEsUnavailable` for all of these situations to signal to plugins that they should handle this error in the same way for every method.

TinaHeiligers replied:
> Personally, I think we should only expose isEsUnavailable for all of these situations to signal to plugins that they should handle this error in the same way for every method.

That makes sense. I'll remove `isNotFoundEsUnavailableError`.

joshdover commented on src/core/server/saved_objects/service/lib/repository.test.js:
```diff
@@ -2221,20 +2228,44 @@ describe('SavedObjectsRepository', () => {
+          elasticsearchClientMock.createSuccessTransportRequestPromise(
+            { found: false },
+            undefined,
+            { 'x-elastic-product': 'Elasticsearch' }
```
Maybe we should include this header as the default in `elasticsearchClientMock.createSuccessTransportRequestPromise()` (or even better might be in the `createApiResponse` function that this function calls) and only override it when needed for these tests. That would better represent the expected typical conditions in our mocks rather than the exceptional case.

TinaHeiligers replied:
I'd prefer to handle that in the follow up issue (that I'm working on right now anyway), as it's going to become the typical conditions when we add the checks to the rest of the SO repository methods.

joshdover commented on src/core/server/saved_objects/service/lib/repository.ts:
```diff
@@ -2197,3 +2227,19 @@ type GetResponseFound<TDocument = unknown> = estypes.GetResponse<TDocument> &
+
+const esXProductMatches = (headers: Record<string, unknown>) => {
+  const esXProductHeaderKey = Object.keys(headers).find(
+    (key: string) => key.toLowerCase() === 'x-elastic-product'
```
It seems this is provided [out of the box by Node.js](https://nodejs.org/api/http.html#http_request_getheader_name), so I think we can remove this extra overhead.

joshdover commented on src/core/server/saved_objects/service/lib/repository.ts:
```diff
@@ -1248,7 +1253,19 @@ export class SavedObjectsRepository {
+      .then((res) => {
+        const indexNotFound = res.statusCode === 404;
+        const esServerSupported = isSupportedEsServer(res.headers);
+        // check if we have the elasticsearch header when doc.found is not true (false, undefined or null) and if we do, ensure it is Elasticsearchq
```
```suggestion
        // check if we have the elasticsearch header when index is not found and if we do, ensure it is Elasticsearch
```

mshustov commented:
ack: going to review tomorrow (04 Aug.)

joshdover commented:
> do you mean the 7.14.1 backport or both backports to 7.14.1 and 7.x (7.15)?

I just mean delaying the backport to 7.14.1

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142745/)
* Commit: b20e9c28afb23765029fcb9ab1b53e0243b88bc1
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-107301/b20e9c28afb23765029fcb9ab1b53e0243b88bc1)
* [Documentation Changes](https://kibana_107301.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/df421f869feb747256bfa329efbbf8c30c603e0e) | [after](https://github.com/elastic/kibana/commit/b20e9c28afb23765029fcb9ab1b53e0243b88bc1) | diff |
| --- | --- | --- | --- |
| `core` | 2363 | 2368 | +5 |

#### API count missing comments
| id | [before](https://github.com/elastic/kibana/commit/df421f869feb747256bfa329efbbf8c30c603e0e) | [after](https://github.com/elastic/kibana/commit/b20e9c28afb23765029fcb9ab1b53e0243b88bc1) | diff |
| --- | --- | --- | --- |
| `core` | 1081 | 1086 | +5 |


</details>

### History
* :green_heart: [Build #142325](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142325/) succeeded 7d00fdd124fb23ec86e99e206fe754ac1655f8fc
* :green_heart: [Build #142074](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142074/) succeeded 5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b
* :green_heart: [Build #142016](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142016/) succeeded e173b778df1e5d5b00c19b1e1c6c6144ff9002d9
* :broken_heart: [Build #141994](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/141994/) failed 98a785fe5a8fac01e138e82d294686021a416e38

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/141994/","number":"141994","commit":"98a785fe5a8fac01e138e82d294686021a416e38"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142016/","number":"142016","commit":"e173b778df1e5d5b00c19b1e1c6c6144ff9002d9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142074/","number":"142074","commit":"5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142325/","number":"142325","commit":"7d00fdd124fb23ec86e99e206fe754ac1655f8fc"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142745/","number":"142745","commit":"b20e9c28afb23765029fcb9ab1b53e0243b88bc1"}]}
PIPELINE-->

TinaHeiligers commented on src/core/server/elasticsearch/client/mocks.ts:
```diff
@@ -141,9 +141,10 @@ export type MockedTransportRequestPromise<T> = TransportRequestPromise<T> & {
   body: T,
-  { statusCode = 200 }: { statusCode?: number } = {}
+  { statusCode = 200 }: { statusCode?: number } = {},
+  headers?: Record<string, any>
```
@joshdover thanks for pointing the type issue out!

TinaHeiligers replied:
[done](https://github.com/elastic/kibana/pull/107301/files#diff-7a19c0f1ba03e2b736d8031dc6c5ea8b746cae37d2d81f357577ec34612dbc12R206)

mshustov commented on src/core/server/saved_objects/service/lib/repository.ts:
```diff
@@ -2087,6 +2104,14 @@ export class SavedObjectsRepository {
     const indexFound = statusCode !== 404;
+
+    // check if we have the elasticsearch header when doc.found is not true (false, undefined or null) and if we do, ensure it is Elasticsearch
+    const esServerSupported = isSupportedEsServer(headers);
```
optional: we repeat the check in several places. It can be extracted in a helper.

TinaHeiligers replied:
True, but other checks are also repeated in a few places. e.g. `indexNotFound`, `!isFoundGetResponse(body)` etc. 

Maybe I'm misunderstanding _what_ should be extracted, we've already extracted the functionality to ensure that the header is present in the response.

mshustov replied:
That's what confuses me. Why in some places do we check `isFoundGetResponse` but in some `indexNotFound`? 
I'd expect we have the same check for all the methods.

TinaHeiligers replied:
> Why in some places do we check isFoundGetResponse but in some indexNotFound?
I'd expect we have the same check for all the methods.

I don't know either. It might be worth trying to clean this up with an overhaul but that's a separate issue.

TinaHeiligers replied:
I'll have to go through the git history for the methods to dig up the possible causes for getting `404` responses.

mshustov replied:
Maybe @rudolf might shed light on the purpose of different conditions here?

rudolf replied:
I can't think of any reason to have different implementations between the methods. When an index isn't found or the document isn't found we should return a 404.

The index not found behaviour is mostly an edge case that I doubt ever happens in practise and when it does there's much bigger problems. In the future I want us to return a 500, when the index isn't there Kibana is broken and in an unrecoverable state.

TinaHeiligers replied:
> When an index isn't found or the document isn't found we should return a 404

I've tried to keep the scope pretty small in this work and only add the header check on the response to `get`, `update` and `delete` when we would otherwise be throwing a 404.

joshdover replied:
> That's what confuses me. Why in some places do we check `isFoundGetResponse` but in some `indexNotFound`?
> I'd expect we have the same check for all the methods.

I believe the reason is that the `update` and `delete` APIs don't return the same `doc.found` property, though they do return a `result` property that may be either `updated` or `deleted` or `not_found`.

I don't think we should change this as part of this PR, it feels out of scope to me.

TinaHeiligers replied:
I've created https://github.com/elastic/kibana/issues/107944 from this comment thread to make sure we don't lose sight of the discussion.

mshustov commented on src/core/server/saved_objects/service/lib/repository.test.js:
```diff
@@ -2221,20 +2228,44 @@ describe('SavedObjectsRepository', () => {
+          elasticsearchClientMock.createSuccessTransportRequestPromise({ found: false })
+        );
+        await expectNotFoundEsUnavailableError(MULTI_NAMESPACE_ISOLATED_TYPE, id);
+        expect(client.get).toHaveBeenCalledTimes(1);
```
nit: I'd delete this assetion. IMO we shouldn't test the internal implementation of `savedObjectsRepository.delete` method.

TinaHeiligers replied:
[done](https://github.com/elastic/kibana/pull/107301/files#diff-8a58e602e79e19512cd392387f38a79ba9782f60d0fc1596a90d5e9e18efb6bdR2253-R2258)

mshustov commented on src/core/server/saved_objects/service/lib/errors.ts:
```diff
@@ -202,4 +202,23 @@ export class SavedObjectsErrorHelpers {
   }
+
+  public static createGenericNotFoundEsUnavailableError(
+    type: string | null = null,
```
under what circumstances can they be `null`?

TinaHeiligers replied:
Good question! This PR only adds the checks to `get`, `update` and `delete`, all of which require a `type` and `id`.

TinaHeiligers replied:
[done](https://github.com/elastic/kibana/pull/107301/files#diff-7a19c0f1ba03e2b736d8031dc6c5ea8b746cae37d2d81f357577ec34612dbc12R206)

TinaHeiligers commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142989/)
* Commit: 0befcbd15e80e39ac03b7211cc322407ed42c3cb
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-107301/0befcbd15e80e39ac03b7211cc322407ed42c3cb)
* [Documentation Changes](https://kibana_107301.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/67634ecfefc39b2b8243c50476efe329ae2bd905) | [after](https://github.com/elastic/kibana/commit/0befcbd15e80e39ac03b7211cc322407ed42c3cb) | diff |
| --- | --- | --- | --- |
| `core` | 2450 | 2455 | +5 |

#### API count missing comments
| id | [before](https://github.com/elastic/kibana/commit/67634ecfefc39b2b8243c50476efe329ae2bd905) | [after](https://github.com/elastic/kibana/commit/0befcbd15e80e39ac03b7211cc322407ed42c3cb) | diff |
| --- | --- | --- | --- |
| `core` | 1165 | 1170 | +5 |


</details>

### History
* :green_heart: [Build #142745](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142745/) succeeded b20e9c28afb23765029fcb9ab1b53e0243b88bc1
* :green_heart: [Build #142325](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142325/) succeeded 7d00fdd124fb23ec86e99e206fe754ac1655f8fc
* :green_heart: [Build #142074](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142074/) succeeded 5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b
* :green_heart: [Build #142016](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142016/) succeeded e173b778df1e5d5b00c19b1e1c6c6144ff9002d9
* :broken_heart: [Build #141994](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/141994/) failed 98a785fe5a8fac01e138e82d294686021a416e38

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/141994/","number":"141994","commit":"98a785fe5a8fac01e138e82d294686021a416e38"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142016/","number":"142016","commit":"e173b778df1e5d5b00c19b1e1c6c6144ff9002d9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142074/","number":"142074","commit":"5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142325/","number":"142325","commit":"7d00fdd124fb23ec86e99e206fe754ac1655f8fc"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142745/","number":"142745","commit":"b20e9c28afb23765029fcb9ab1b53e0243b88bc1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142989/","number":"142989","commit":"0befcbd15e80e39ac03b7211cc322407ed42c3cb"}]}
PIPELINE-->

TinaHeiligers commented on src/core/server/saved_objects/service/lib/repository.ts:
```diff
@@ -2197,3 +2222,12 @@ type GetResponseFound<TDocument = unknown> = estypes.GetResponse<TDocument> &
+ * @returns boolean
+ */
+const isSupportedEsServer = (headers: Record<string, unknown> | null) => {
+  return !!headers && headers['x-elastic-product'] === 'Elasticsearch';
```
@mshustov Agreed. However, I wanted to minimize changes to services in this PR so that the risk is lower for backporting to a patch release. The alerting team really needs this workaround as soon as possible.

mshustov replied:
> We can revisit in the follow up PR that implements the checks in other SO repository methods.

Could you add a comment to move it to ES service? 

> I'm also not sure exactly where to move it to 🤔

in a dedicated file is okay

TinaHeiligers replied:
I'm making other changes to the PR and have moved the method, as requested.

TinaHeiligers commented:
@joshdover which question did you give the thumbs up for in https://github.com/elastic/kibana/pull/107301#discussion_r682480245? i.e.

> Maybe we should pass an error reason to decorateEsUnavailableError explicitly?

or 

> Will we add such a helper for every possible error response?

I'm not sure which way we want to go eventually, as long as we provide enough info to easily figure out the root cause of an error when it's thrown.

TinaHeiligers commented on src/core/server/saved_objects/service/lib/repository.ts:
```diff
@@ -2197,3 +2222,12 @@ type GetResponseFound<TDocument = unknown> = estypes.GetResponse<TDocument> &
+ * @param headers Response headers
+ * @returns boolean
+ */
+const isSupportedEsServer = (headers: Record<string, unknown> | null) => {
```
[done](https://github.com/elastic/kibana/pull/107301/files#diff-01c073903639e29be8562ed0de79fae4084a953304672d1e6bdb3fd7fef06f11R2231)

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143065/)
* Commit: 9ab33bb088fd953ed63d70d6ca4735d68c3d25ea
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types and Public API Docs](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143065/execution/node/428/log)

### History
* :green_heart: [Build #142989](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142989/) succeeded 0befcbd15e80e39ac03b7211cc322407ed42c3cb
* :green_heart: [Build #142745](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142745/) succeeded b20e9c28afb23765029fcb9ab1b53e0243b88bc1
* :green_heart: [Build #142325](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142325/) succeeded 7d00fdd124fb23ec86e99e206fe754ac1655f8fc
* :green_heart: [Build #142074](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142074/) succeeded 5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b
* :green_heart: [Build #142016](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142016/) succeeded e173b778df1e5d5b00c19b1e1c6c6144ff9002d9

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142016/","number":"142016","commit":"e173b778df1e5d5b00c19b1e1c6c6144ff9002d9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142074/","number":"142074","commit":"5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142325/","number":"142325","commit":"7d00fdd124fb23ec86e99e206fe754ac1655f8fc"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142745/","number":"142745","commit":"b20e9c28afb23765029fcb9ab1b53e0243b88bc1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142989/","number":"142989","commit":"0befcbd15e80e39ac03b7211cc322407ed42c3cb"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143065/","number":"143065","commit":"9ab33bb088fd953ed63d70d6ca4735d68c3d25ea"}]}
PIPELINE-->

TinaHeiligers commented:
@joshdover @mshustov I've addressed the nits and comments from your reviews. I'm not sure if they were explicitly change requests.
I haven't set auto-merge or auto-backport, so approval from a core team member won't automatically merge this and you may give it the LGTM if it's ok for now.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143256/)
* Commit: 987c94198c70c7f2b8e35fade1f7aa0420a18392
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types and Public API Docs](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143256/execution/node/428/log)

### History
* :broken_heart: [Build #143065](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143065/) failed 9ab33bb088fd953ed63d70d6ca4735d68c3d25ea
* :green_heart: [Build #142989](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142989/) succeeded 0befcbd15e80e39ac03b7211cc322407ed42c3cb
* :green_heart: [Build #142745](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142745/) succeeded b20e9c28afb23765029fcb9ab1b53e0243b88bc1
* :green_heart: [Build #142325](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142325/) succeeded 7d00fdd124fb23ec86e99e206fe754ac1655f8fc
* :green_heart: [Build #142074](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142074/) succeeded 5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142074/","number":"142074","commit":"5614cda7ba4dda9679fcc4ec747b4bd7a5d35d5b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142325/","number":"142325","commit":"7d00fdd124fb23ec86e99e206fe754ac1655f8fc"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142745/","number":"142745","commit":"b20e9c28afb23765029fcb9ab1b53e0243b88bc1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142989/","number":"142989","commit":"0befcbd15e80e39ac03b7211cc322407ed42c3cb"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143065/","number":"143065","commit":"9ab33bb088fd953ed63d70d6ca4735d68c3d25ea"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143256/","number":"143256","commit":"987c94198c70c7f2b8e35fade1f7aa0420a18392"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143299/)
* Commit: 35cf9d0f46b6def889044e698fa21d0cc73735f9
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Check Doc API Changes](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143299/execution/node/401/log)

### History
* :broken_heart: [Build #143256](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143256/) failed 987c94198c70c7f2b8e35fade1f7aa0420a18392
* :broken_heart: [Build #143065](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143065/) failed 9ab33bb088fd953ed63d70d6ca4735d68c3d25ea
* :green_heart: [Build #142989](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142989/) succeeded 0befcbd15e80e39ac03b7211cc322407ed42c3cb
* :green_heart: [Build #142745](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142745/) succeeded b20e9c28afb23765029fcb9ab1b53e0243b88bc1
* :green_heart: [Build #142325](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142325/) succeeded 7d00fdd124fb23ec86e99e206fe754ac1655f8fc

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142325/","number":"142325","commit":"7d00fdd124fb23ec86e99e206fe754ac1655f8fc"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142745/","number":"142745","commit":"b20e9c28afb23765029fcb9ab1b53e0243b88bc1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142989/","number":"142989","commit":"0befcbd15e80e39ac03b7211cc322407ed42c3cb"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143065/","number":"143065","commit":"9ab33bb088fd953ed63d70d6ca4735d68c3d25ea"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143256/","number":"143256","commit":"987c94198c70c7f2b8e35fade1f7aa0420a18392"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143299/","number":"143299","commit":"35cf9d0f46b6def889044e698fa21d0cc73735f9"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143334/)
* Commit: 81d7571dac7ebfa908fcd0e56ecc4ddea1f301ae
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-107301/81d7571dac7ebfa908fcd0e56ecc4ddea1f301ae)
* [Documentation Changes](https://kibana_107301.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/ca6182cd20d88f7717f6495d52190a204c537152) | [after](https://github.com/elastic/kibana/commit/81d7571dac7ebfa908fcd0e56ecc4ddea1f301ae) | diff |
| --- | --- | --- | --- |
| `core` | 2450 | 2455 | +5 |

#### API count missing comments
| id | [before](https://github.com/elastic/kibana/commit/ca6182cd20d88f7717f6495d52190a204c537152) | [after](https://github.com/elastic/kibana/commit/81d7571dac7ebfa908fcd0e56ecc4ddea1f301ae) | diff |
| --- | --- | --- | --- |
| `core` | 1165 | 1170 | +5 |


</details>

### History
* :broken_heart: [Build #143299](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143299/) failed 35cf9d0f46b6def889044e698fa21d0cc73735f9
* :broken_heart: [Build #143256](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143256/) failed 987c94198c70c7f2b8e35fade1f7aa0420a18392
* :broken_heart: [Build #143065](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143065/) failed 9ab33bb088fd953ed63d70d6ca4735d68c3d25ea
* :green_heart: [Build #142989](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142989/) succeeded 0befcbd15e80e39ac03b7211cc322407ed42c3cb
* :green_heart: [Build #142745](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142745/) succeeded b20e9c28afb23765029fcb9ab1b53e0243b88bc1

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142745/","number":"142745","commit":"b20e9c28afb23765029fcb9ab1b53e0243b88bc1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142989/","number":"142989","commit":"0befcbd15e80e39ac03b7211cc322407ed42c3cb"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143065/","number":"143065","commit":"9ab33bb088fd953ed63d70d6ca4735d68c3d25ea"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143256/","number":"143256","commit":"987c94198c70c7f2b8e35fade1f7aa0420a18392"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143299/","number":"143299","commit":"35cf9d0f46b6def889044e698fa21d0cc73735f9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143334/","number":"143334","commit":"81d7571dac7ebfa908fcd0e56ecc4ddea1f301ae"}]}
PIPELINE-->

TinaHeiligers commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143530/)
* Commit: 9cbbb57a3f61132331c66ceb95597038720ab637
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-107301/9cbbb57a3f61132331c66ceb95597038720ab637)
* [Documentation Changes](https://kibana_107301.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/b16d20b50aa72dac9a1dbc1f7e3e57b5c24f77ef) | [after](https://github.com/elastic/kibana/commit/9cbbb57a3f61132331c66ceb95597038720ab637) | diff |
| --- | --- | --- | --- |
| `core` | 2455 | 2460 | +5 |

#### API count missing comments
| id | [before](https://github.com/elastic/kibana/commit/b16d20b50aa72dac9a1dbc1f7e3e57b5c24f77ef) | [after](https://github.com/elastic/kibana/commit/9cbbb57a3f61132331c66ceb95597038720ab637) | diff |
| --- | --- | --- | --- |
| `core` | 1166 | 1171 | +5 |


</details>

### History
* :green_heart: [Build #143334](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143334/) succeeded 81d7571dac7ebfa908fcd0e56ecc4ddea1f301ae
* :broken_heart: [Build #143299](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143299/) failed 35cf9d0f46b6def889044e698fa21d0cc73735f9
* :broken_heart: [Build #143256](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143256/) failed 987c94198c70c7f2b8e35fade1f7aa0420a18392
* :broken_heart: [Build #143065](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143065/) failed 9ab33bb088fd953ed63d70d6ca4735d68c3d25ea
* :green_heart: [Build #142989](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142989/) succeeded 0befcbd15e80e39ac03b7211cc322407ed42c3cb

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/142989/","number":"142989","commit":"0befcbd15e80e39ac03b7211cc322407ed42c3cb"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143065/","number":"143065","commit":"9ab33bb088fd953ed63d70d6ca4735d68c3d25ea"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143256/","number":"143256","commit":"987c94198c70c7f2b8e35fade1f7aa0420a18392"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143299/","number":"143299","commit":"35cf9d0f46b6def889044e698fa21d0cc73735f9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143334/","number":"143334","commit":"81d7571dac7ebfa908fcd0e56ecc4ddea1f301ae"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143530/","number":"143530","commit":"9cbbb57a3f61132331c66ceb95597038720ab637"}]}
PIPELINE-->

joshdover commented:
I'm ✅ on this PR, only pending the discussion in https://github.com/elastic/kibana/pull/107301#discussion_r685202829

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143817/)
* Commit: 2af49d32a480ec9304336b53b170fa854cadd461
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-107301/2af49d32a480ec9304336b53b170fa854cadd461)
* [Documentation Changes](https://kibana_107301.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/b16d20b50aa72dac9a1dbc1f7e3e57b5c24f77ef) | [after](https://github.com/elastic/kibana/commit/2af49d32a480ec9304336b53b170fa854cadd461) | diff |
| --- | --- | --- | --- |
| `core` | 2455 | 2458 | +3 |

#### API count missing comments
| id | [before](https://github.com/elastic/kibana/commit/b16d20b50aa72dac9a1dbc1f7e3e57b5c24f77ef) | [after](https://github.com/elastic/kibana/commit/2af49d32a480ec9304336b53b170fa854cadd461) | diff |
| --- | --- | --- | --- |
| `core` | 1166 | 1169 | +3 |


</details>

### History
* :green_heart: [Build #143530](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143530/) succeeded 9cbbb57a3f61132331c66ceb95597038720ab637
* :green_heart: [Build #143334](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143334/) succeeded 81d7571dac7ebfa908fcd0e56ecc4ddea1f301ae
* :broken_heart: [Build #143299](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143299/) failed 35cf9d0f46b6def889044e698fa21d0cc73735f9
* :broken_heart: [Build #143256](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143256/) failed 987c94198c70c7f2b8e35fade1f7aa0420a18392
* :broken_heart: [Build #143065](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143065/) failed 9ab33bb088fd953ed63d70d6ca4735d68c3d25ea

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143065/","number":"143065","commit":"9ab33bb088fd953ed63d70d6ca4735d68c3d25ea"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143256/","number":"143256","commit":"987c94198c70c7f2b8e35fade1f7aa0420a18392"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143299/","number":"143299","commit":"35cf9d0f46b6def889044e698fa21d0cc73735f9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143334/","number":"143334","commit":"81d7571dac7ebfa908fcd0e56ecc4ddea1f301ae"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143530/","number":"143530","commit":"9cbbb57a3f61132331c66ceb95597038720ab637"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143817/","number":"143817","commit":"2af49d32a480ec9304336b53b170fa854cadd461"}]}
PIPELINE-->

TinaHeiligers commented:
@elasticmachine run elasticsearch-ci/docs

jbudz commented:
@elasticmachine run elasticsearch-ci/docs

TinaHeiligers commented:
jenkins test this please

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143899/)
* Commit: 2af49d32a480ec9304336b53b170fa854cadd461
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-107301/2af49d32a480ec9304336b53b170fa854cadd461)
* This comment will update when the build is complete

### Failed CI Steps
* [Execute kibana-ciGroup7](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143899/execution/node/821/log)
* [Execute kibana-ciGroup7](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143899/execution/node/826/log)

### History
* :green_heart: [Build #143817](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143817/) succeeded 2af49d32a480ec9304336b53b170fa854cadd461
* :green_heart: [Build #143530](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143530/) succeeded 9cbbb57a3f61132331c66ceb95597038720ab637
* :green_heart: [Build #143334](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143334/) succeeded 81d7571dac7ebfa908fcd0e56ecc4ddea1f301ae
* :broken_heart: [Build #143299](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143299/) failed 35cf9d0f46b6def889044e698fa21d0cc73735f9
* :broken_heart: [Build #143256](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143256/) failed 987c94198c70c7f2b8e35fade1f7aa0420a18392

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143256/","number":"143256","commit":"987c94198c70c7f2b8e35fade1f7aa0420a18392"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143299/","number":"143299","commit":"35cf9d0f46b6def889044e698fa21d0cc73735f9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143334/","number":"143334","commit":"81d7571dac7ebfa908fcd0e56ecc4ddea1f301ae"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143530/","number":"143530","commit":"9cbbb57a3f61132331c66ceb95597038720ab637"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143817/","number":"143817","commit":"2af49d32a480ec9304336b53b170fa854cadd461"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143899/","number":"143899","commit":"2af49d32a480ec9304336b53b170fa854cadd461"}]}
PIPELINE-->

TinaHeiligers commented:
@elasticmachine merge upstream

TinaHeiligers commented:
**Summary of PR status:**
Initial implementation of extra check to make sure we get a response from elasticsearch before throwing 404’s. This handles the immediate need for the Alerting team, where they need to make sure we get a response from elasticsearch before deleting an alert that isn’t found.

@dover @mshustov @rudolf I’ve created the following issues to follow up on from this work:
1. Implement similar checks in all the methods (https://github.com/elastic/kibana/issues/107343)
2. Add integration test as a smoke screen to make sure the behavior is as expected. (https://github.com/elastic/kibana/issues/107246)
3. Improve on the way we interpret 404 responses and change to throwing 500 errors when the index isn’t found. (https://github.com/elastic/kibana/issues/107944)

Please let me know if I’ve missed something or if there are remaining issues to tackle before moving forward with the PR.

I’m hoping we can a fix in before 7.15 FF but, depending on priorities, don't know if there's enough time to get through all the follow up issues before FF.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143913/)
* Commit: 2f76296dbb161a186ae1347e72c838741e3f74fb
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-107301/2f76296dbb161a186ae1347e72c838741e3f74fb)
* [Documentation Changes](https://kibana_107301.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/b4dfe9aeb166a4700c354d7b1b17bb9d730d95be) | [after](https://github.com/elastic/kibana/commit/2f76296dbb161a186ae1347e72c838741e3f74fb) | diff |
| --- | --- | --- | --- |
| `core` | 2455 | 2458 | +3 |

#### API count missing comments
| id | [before](https://github.com/elastic/kibana/commit/b4dfe9aeb166a4700c354d7b1b17bb9d730d95be) | [after](https://github.com/elastic/kibana/commit/2f76296dbb161a186ae1347e72c838741e3f74fb) | diff |
| --- | --- | --- | --- |
| `core` | 1166 | 1169 | +3 |


</details>

### History
* :broken_heart: [Build #143899](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143899/) failed 2af49d32a480ec9304336b53b170fa854cadd461
* :green_heart: [Build #143817](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143817/) succeeded 2af49d32a480ec9304336b53b170fa854cadd461
* :green_heart: [Build #143530](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143530/) succeeded 9cbbb57a3f61132331c66ceb95597038720ab637
* :green_heart: [Build #143334](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143334/) succeeded 81d7571dac7ebfa908fcd0e56ecc4ddea1f301ae
* :broken_heart: [Build #143299](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143299/) failed 35cf9d0f46b6def889044e698fa21d0cc73735f9

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143299/","number":"143299","commit":"35cf9d0f46b6def889044e698fa21d0cc73735f9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143334/","number":"143334","commit":"81d7571dac7ebfa908fcd0e56ecc4ddea1f301ae"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143530/","number":"143530","commit":"9cbbb57a3f61132331c66ceb95597038720ab637"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143817/","number":"143817","commit":"2af49d32a480ec9304336b53b170fa854cadd461"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143899/","number":"143899","commit":"2af49d32a480ec9304336b53b170fa854cadd461"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/143913/","number":"143913","commit":"2f76296dbb161a186ae1347e72c838741e3f74fb"}]}
PIPELINE-->

TinaHeiligers commented:
Note to self: backport manually to 7.14.1.

kibanamachine commented:
## 💚 Backport successful

| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [7.x](https://github.com/elastic/kibana/pull/108037)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/108037">](https://github.com/elastic/kibana/pull/108037) |

This backport PR will be merged automatically after passing CI.

ymao1 commented:
@TinaHeiligers Was this backported to 7.14? The tag says 7.14 but I can't find a backport PR. Wondering because we are seeing a saved object not found error for a rule in 7.14.1 in a cluster that seems to have some proxy connectivity issues.

TinaHeiligers commented:
@ymao1 no, it wasn't backported to 7.14.1.
